ESPS/general/src/mat2fea/mat2fea.1
.\" Copyright (c) 1991 Entropic Research Laboratory, Inc.; All rights reserved
.\" @(#)mat2fea.1	1.2 7/23/91 ERL
.ds ]W (c) 1991 Entropic Research Laboratory, Inc.
.TH MAT2FEA 1\-ESPS 7/10/91
.sh NAME
.nf
mat2fea \- convert Matlab .mat file data to FEA file
.SH SYNOPSIS
.B mat2fea
[
.BI \-c " comment_text"
] [
.BI \-x " debug_level"
] [
.BI \-m
] 
.BI \-f " variable_name"
.I infile.mat outfile.fea
.SH DESCRIPTION
This program reads the Matlab matrix variable \fIvariable_name\fP from 
the Matlab data file \fIinfile.mat\fP, creates the ESPS FEA file 
\fIoutfile.fea\fP,
and writes the variable data 
to the FEA field \fIvariable_name\fP.
The variable is taken as a \fIM x N\fP matrix.
The default behavior is to create a field of size \fIN\fP in the output
file and write row \fIn\fP to record \fIn\fP so that \fIoutfile.fea\fP
has \fIN\fP records.  If \fB\-m\fP is specified, the output field is
created as a \fIM x N\fP matrix and the data is written to a single record of 
\fIoutfile.fea\fP.  Matlab string variables (ASCII data) are not supported.
.SH OPTIONS
.PP
The following options are supported:
.TP
.BI \-x " debug_level" "\fR [0]\fP"
Increasing values provide more detailed commentary on program's progress.
.TP
.BI \-m 
Write data to single record output file as a matrix. Default behavior is
is to write one record for each row of the matrix.
.TP
.BI \-f " field_name" 
Name of variable in the input file from which data is read - this field
must be provided.
.SH ESPS PARAMETERS
.PP
No parameter file is read.
.SH ESPS COMMON
.PP
No common processing is performed.
.SH ESPS HEADERS
The command line and comment string, if provided, are added to the
output header as comments.  The output field \fIvariable_name\fP
has data type DOUBLE or DOUBLE_CPLX, depending on the input data.
.SH BUGS
Sometimes has trouble reading information about the variables when 
\fIinfile.mat\fP contains more than one
variable.
.SH "SEE ALSO"
fea2mat(1\-ESPS), FEA(5\-ESPS)
.SH AUTHOR
Program and man page by Bill Byrne.



ESPS/general/src/mat2fea/mat2fea.c
/*
 * This material contains unpublished, proprietary software of 
 * Entropic Research Laboratory, Inc. Any reproduction, distribution, 
 * or publication of this work must be authorized in writing by Entropic 
 * Research Laboratory, Inc., and must bear the notice: 
 *
 *    "Copyright (c) 1990-1991  Entropic Research Laboratory, Inc. 
 *                   All rights reserved"
 *
 * The copyright notice above does not evidence any actual or intended 
 * publication of this source code.     
 *
 * Written by:  Bill Byrne
 * Checked by:
 * Revised by:
 *
 * Brief description: read variable from Matlab .mat file and write to FEA file
 *
 */

static char *sccs_id = "@(#)mat2fea.c	1.2	10/20/93	ERL";

#include <stdio.h>
#include <esps/esps.h>
#include <esps/fea.h>

#define Fprintf (void)fprintf
#define SYNTAX {\
Fprintf(stderr, "%s: import MATLAB files into ESPS FEA files.\n", PROG); \
Fprintf(stderr, "Usage:  %s [-c \"comment text\"] [-x debug_level] [-m] -f variable infile.mat outfile.fea \n", PROG); \
exit(1); }

typedef struct {
     long type;   /* type */
     long mrows;  /* row dimension */
     long ncols;  /* column dimension */
     long imagf;  /* flag indicating imag part */
     long namlen; /* name length (including NULL) */
} Fmatrix;


int debug_level=0;
char *PROG="mat2fea";
char *VERSION="1.2";
char *DATE="10/20/93";

char *get_cmd_line();
char *marg_index();
#ifndef DEC_ALPHA
char *strcpy();
char *calloc();
#endif

main( argc, argv)
int argc;
char **argv;
{
    extern char *optarg;
    extern int optind;
    char opt;

    struct header *outhd;
    struct fea_data *outrec;
    char *comment=NULL;

    FILE *fpin, *fpout; 
    char *fnameout;

    Fmatrix x, xread;

    char *fname=NULL;
    char *targ_fname;
    long dim[2];
    short rank;
    long size;
    short type;

    char *cptr;
    double *dd, **mdd;
    double_cplx *cdd, **cmdd;

    double *rmptr, *imptr;

    int arrsize;
    int cnt;
    int fflag=0;
    int mflag=0;
    int field_found=0;
    int maxfnamelen=0;

    int m, n;

    while ( (opt = getopt(argc, argv, "x:c:f:m")) != EOF ) 
	switch(opt) {
	    case 'x': {
		debug_level=atoi(optarg);
		break;
	    }
	    case 'f': {
		fflag++;
		targ_fname = optarg;
		break;
	    }
	    case 'c': {
		comment = optarg;
		break;
	    }
	    case 'm': {
		mflag=1;
		break;
	    }
	}
    if (!fflag) {
	Fprintf(stderr, "%s: must specify field name.\n", PROG);
	SYNTAX;
    }

    if ( (optind+2) != argc) {
	Fprintf(stderr, "%s: must specify an input and an output file.", PROG);
	SYNTAX;
    }
    
    if ( (fpin=fopen(argv[optind++], "r")) == NULL ) {
	Fprintf(stderr, "%s: can't open input file %s.\n", argv[optind-1], PROG);
	exit(1);
    }

    field_found = 0;
    while ( fread((char *)&x, sizeof(Fmatrix), (int)1, fpin ) == 1 && !field_found ) {

	if (x.mrows<=0 || x.ncols <=0) {
	    Fprintf(stderr, "trouble reading input file - bad header structure.\n");
	    exit(1);
	}
	if (x.namlen > maxfnamelen) {
	    maxfnamelen = x.namlen;
	    if (fname != NULL)
		free(fname);
	    fname = (char *) calloc( (unsigned)  maxfnamelen, sizeof(char));
	    spsassert( fname!= NULL, "can't allocate memory for field name.");
	}

	if ( fread((char *) fname, sizeof(char), (int)x.namlen, fpin) != x.namlen) 
	    break;

	if (debug_level) {
	    if (x.imagf)
		Fprintf(stderr, "%s: found field %s %dx%d IMAG\n", PROG, fname, x.mrows, x.ncols);
	    else
		Fprintf(stderr, "%s: found field %s %dx%d REAL\n", PROG, fname, x.mrows, x.ncols);
	}    

	if ( strcmp( fname, targ_fname ) == 0 ) {
	    xread = x;
	    spsassert( !(xread.type & 1), "matlab text fields not supported.");
	    field_found=1;
	    arrsize = xread.mrows * xread.ncols;
	    rmptr = (double *) calloc( (unsigned) arrsize, sizeof(double));
	    spsassert( rmptr!=NULL, "can't allocate output matrix memory.");
	    cnt = fread((char *)rmptr, sizeof(double), (int) arrsize, fpin);
	    spsassert(cnt == arrsize, "short read from input file.");
	    if (xread.imagf) {
		imptr = (double *) calloc( (unsigned) arrsize, sizeof(double));
		spsassert( imptr!=NULL, "can't allocate output matrix memory.");
		cnt = fread((char *)imptr, sizeof(double), (int) arrsize, fpin);
		spsassert(cnt == arrsize, "short read from input file.");
	    }
	}
    }
    (void)fclose(fpin);
    if (!field_found) {
	Fprintf(stderr, "%s: couldn't field field %s.\n", PROG, targ_fname);
	exit(1);
    }

    fnameout = eopen( PROG, argv[optind++], "w", FT_FEA, NONE, &outhd, &fpout);
    if (debug_level)
	Fprintf(stderr, "%s: writing to FEA file %s\n", PROG, fnameout);

    outhd = new_header(FT_FEA);
    (void) strcpy(outhd->common.prog, PROG);
    (void) strcpy(outhd->common.vers, VERSION);
    (void) strcpy(outhd->common.progdate, DATE);

    add_comment( outhd, get_cmd_line(argc, argv));
    if (comment != NULL )
	add_comment( outhd, comment);

    if (mflag) {
	dim[0] = xread.mrows;
	dim[1] = xread.ncols;
	size = dim[0] * dim[1];
	rank = 2;
    } else {
	if (xread.ncols==1)
	    rank = 0;
	else
	    rank = 1;
	size = xread.ncols;
    }
    if (xread.imagf)
	type = DOUBLE_CPLX;
    else
	type = DOUBLE;

    if ( add_fea_fld( fname, size, rank, dim, type, (char **)NULL, outhd) != 0 ) {
	Fprintf(stderr, "%s: unable to add output field.\n");
	exit(1);
    }
    
    write_header( outhd, fpout);
    outrec = allo_fea_rec(outhd);

    if (mflag) {

	if ( xread.imagf ) {
	    cptr = (char *) get_fea_ptr( outrec, fname, outhd);
	    cmdd = (double_cplx **) marg_index( (char *) cptr, (short)2, dim, DOUBLE_CPLX, 0); 
	    spsassert( cmdd!=NULL, "can't allocate output record memory.");
	}  else {
	    cptr = (char *) get_fea_ptr( outrec, fname, outhd);
	    mdd = (double **) marg_index( (char *) cptr, (short)2, dim, DOUBLE, 0); 
	    spsassert( mdd!=NULL, "can't allocate output record memory.");
	}

	for (m=0; m<xread.mrows; m++)
	    for (n=0; n<xread.ncols; n++) {
		cnt = n * xread.mrows + m;
		if ( xread.imagf ) {
		    cmdd[m][n].real = rmptr[cnt];
		    cmdd[m][n].imag = imptr[cnt];
		} else 
		    mdd[m][n] = rmptr[cnt];
	    } 

	put_fea_rec( outrec, outhd, fpout);

    } else {

	if ( xread.imagf ) {
	    cdd = (double_cplx *) get_fea_ptr( outrec, fname, outhd);
	    spsassert( cdd!=NULL, "can't allocate output record memory.");
	} else {
	    dd = (double *) get_fea_ptr( outrec, fname, outhd);
	    spsassert( dd!=NULL, "can't allocate output record memory.");
	} 
	
	for (m=0; m<xread.mrows; m++) {
	    for (n=0; n<xread.ncols; n++) {
		cnt = n * xread.mrows + m;
		if (xread.imagf) {
		    cdd[n].real = rmptr[cnt];
		    cdd[n].imag = imptr[cnt];
		} else
		    dd[n] = rmptr[cnt];
	    }		
	    put_fea_rec( outrec, outhd, fpout);
	}
    }
    
    (void)fclose(fpout);
}


